### DISCUSSION TITLE
Theia Settings Menu Modification

### DISCUSSION THREAD

**Discussion by maheshrajrp** [eclipse-theia/theia#9857] [General] [answered]
Hi All,
I wish to edit the settings menu available in theia. I wish add a new menu item within that upon clicking that it should display a dialog-widget. I'm new to theia. Any help would be appreciated. Thanks in advance. 

![image](https://user-images.githubusercontent.com/80001447/128443214-8ca3ab5c-66b1-44fe-9a4b-0d568a46cc7b.png)

**Comment by msujew** [comment, marked as answer]
Hi @maheshrajrp,

You can add new menu items by using the `CommonMenus.SETTINGS_OPEN` menu path like this in your `MenuContribution`:

```ts
menuRegistry.registerMenuAction(CommonMenus.SETTINGS_OPEN, {
    commandId: DIALOG_COMMAND.id
});
```

Independently of that, you need to register a command that corresponds to the menu item in your `CommandContribution`:

```ts
const DIALOG_COMMAND = {
    id: 'open-custom-dialog-widget',
    label: 'Open dialog'
}
...
commandRegistry.registerCommand(DIALOG_COMMAND , {
    execute: () => openDialogWidget()
});
```

The simplest implementation of some `openDialogWidget()` method would look like this:
```ts
openDialogWidget(): void {
    new ConfirmDialog({
        title: 'Open dialog',
        msg: 'Hi, this is a simple confirm dialog'
    }).open();
}
```

Of course, you can create any kind of widget in there that you like.

**Comment by maheshrajrp** [comment]
Hi @msujew ,
Thanks for your time. Your comment was easy to understand. Also, it would be great if you could assist with a way to access this menu. Thus, I would be able to link this menu item to the dropdown menu. 
Thanks.

![image](https://user-images.githubusercontent.com/80001447/128521341-ea4c6d44-7f12-4f98-8c95-68d618e28023.png)

**Reply by msujew** [reply]
Do you mean specifically the color themes, or the quick input in general? The quick input is responsible for all kinds of dropdown menus that look like the one in your image. It quiet easy to use:

```ts
const items: QuickPickItem[] = [
    { label: '1', execute: () => console.log('1') },
    { label: '2', execute: () => console.log('2') },
    { label: '3', execute: () => console.log('3') }
];
this.quickInputService.showQuickPick(items, {
    placeholder: 'Select 1, 2 or 3'
});
```

Of course you need to inject it first, like this:

```ts
@inject(QuickInputService)
protected quickInputService: QuickInputService;
```

Perhaps you need to update to Theia version `1.16.0` first, because the quick input API changed a bit with that update.

**Reply by maheshrajrp** [reply]
Hi @msujew ,
I'm having trouble finding menuRegistry in menu-contribution, do I need to inject it ? If yes can you help with the class.
And Thanks for the quick reply.

**Reply by maheshrajrp** [reply]
I've found the menuRegistry. Thanks !!

**Reply by msujew** [reply]
Ah, I see, probably should've explained that a bit better. Take a look at the following example:

```ts
const DIALOG_COMMAND = {
    id: 'open-custom-dialog-widget',
    label: 'Open dialog'
}

export class CustomContribution implements CommandContribution, MenuContribution {
    registerCommands(commandRegistry: CommandRegistry) {
        commandRegistry.registerCommand(DIALOG_COMMAND , {
            execute: () => openDialogWidget()
        });
    }
    
    registerMenus(menuRegistry: MenuRegistry) {
        menuRegistry.registerMenuAction(CommonMenus.SETTINGS_OPEN, {
            commandId: DIALOG_COMMAND.id
        });
    }
}
```

Basically, the `CommandContribution` interface tries to implement `registerCommands` and the `MenuContribution` tries implementing `menuRegistry`. You need to bind everything in the end like this:

```ts
bind(CustomContribution).toSelf().inSingletonscope();
bind(CommandContribution).toService(CustomContribution);
bind(MenuContribution).toService(CustomContribution);
```

**Reply by maheshrajrp** [reply]
Hi @msujew ,
You are awesome. I've now implemented them. I'm yet to understand how these binds work and how do you figure out which bind to use like there are muliple type bind.toSelf, bind.toService and what is inSingletonscope. I know I'm pushing but it would great if you could provide somelinks that can help me understand them. And again Thanks !! You Are Awesome... !!

**Reply by msujew** [reply]
Dependency injection can be a tricky thing, so I can understand you quite well in that regard.

Theia uses [`inversify`](https://inversify.io/) for dependency injection. [Here](https://theia-ide.org/docs/services_and_contributions/)'s a short documentation on that topic. So in short:

* `bind(CustomContribution).toSelf()` will bind every injection of `CustomContribution` to itself. Meaning, every call of `@inject(CustomContribution)` will inject an instance of `CustomContribution`. `inSingletonScope()` signals that the scope of the binding should be changed to `singleton`. This means, that every `@inject(CustomContribution)` will actually inject the same instance. Compare this to `inTransientScope()` which will always create a new instance of `CustomContribution` for every `@inject`.
* `toService` does something quite interesting. So basically, in Theia you can add different kinds of `CommandContribution`, `MenuContributions`, etc. In Theia's own codebase, all of these bound instances will be collected by a generic `ContributionProvider`. For example, a service that needs __all__ bound `CommandContribution` elements can simply inject the `ContributionProvider<CommandContribution>`. You need the `toService` binding, so that your services are included in that `ContributionProvider`.

I can't give a general recommendation, on what to use at what times. Please take a look at the Theia codebase itself, to see different usages of the inversify framework.

**Reply by maheshrajrp** [reply]
Sure, Thanks @msujew . You've just saved me days of research.

**Reply by maheshrajrp** [reply]
Hi @msujew ,

Can you guide me in adding errors and warning to this Problems space. 

![image](https://user-images.githubusercontent.com/80001447/128658896-c47893d0-032f-46f3-98dd-73c9128eee90.png)

Thanks !

**Reply by msujew** [reply]
@maheshrajrp You need to add extensions for different languages first (Java, JS, TS, etc.). Having invalid files (syntax errors, or something similiar) in your workspace will then populate the problems view. Or what exactly are you refering to?

**Reply by maheshrajrp** [reply]
Hi @msujew ,
So, I'm currently working on a project where they have built a layer above theia where process like build happens and errors and warnings are throw. So, I'm trying link those errors and warnings back to the theia layer. I would like know if I can link our customs functions to throw errors that can be linked with that Problems Space Icon.
